Source/WebCore/dom/Element.cpp

```diff
@@ -2735,232 +2735,307 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
-            element->setEffectiveTextDirection(effectiveDirection);
-        it.traverseNext();
     }
+
+    return std::nullopt;
 }
 
-void Element::adjustDirectionalityIfNeededAfterChildrenChanged(Element* beforeChange, ChildChange::Type changeType)
+static std::optional<TextDirection> computeTextFormControlElementTextDirection(const HTMLTextFormControlElement& textFormControlElement)
 {
-    // FIXME: This function looks suspicious.
+    if (!textFormControlElement.dirAutoUsesValue())
+        return computeElementContainedTextAutoDirection(textFormControlElement);
 
-    if (!selfOrPrecedingNodesAffectDirAuto())
-        return;
+    // Specs: The directionality of the auto-directionality form-associated
+    // element is calculated from its value() text.
+    // Specs: If element's value is not the empty string, then return 'ltr'.
+    if (auto value = textFormControlElement.value(); !value.isEmpty())
+        return computeTextDirectionFromText(value).value_or(TextDirection::LTR);
+    return std::nullopt;
+}
 
-    RefPtr<Node> oldMarkedNode;
-    if (beforeChange)
-        oldMarkedNode = changeType == ChildChange::Type::ElementInserted ? ElementTraversal::nextSibling(*beforeChange) : beforeChange->nextSibling();
+static void updateElementAndShadowEffectiveTextDirection(Element& element, std::optional<TextDirection> direction, Element* exclude = nullptr)
+{
+    bool usesEffectiveTextDirection = !!direction;
+    auto effectiveDirection = direction.value_or(TextDirection::LTR);
 
-    while (oldMarkedNode && elementAffectsDirectionality(*oldMarkedNode))
-        oldMarkedNode = oldMarkedNode->nextSibling();
-    if (oldMarkedNode)
-        setHasDirAutoFlagRecursively(oldMarkedNode.get(), false);
+    Style::PseudoClassChangeInvalidation styleInvalidation(element, CSSSelector::PseudoClass::Dir, Style::PseudoClassChangeInvalidation::AnyValue);
+    element.setUsesEffectiveTextDirection(usesEffectiveTextDirection);
+    element.setEffectiveTextDirection(effectiveDirection);
 
-    for (Ref elementToAdjust : lineageOfType<HTMLElement>(*this)) {
-        if (elementAffectsDirectionality(elementToAdjust)) {
-            ASSERT(elementToAdjust->hasDirectionAuto());
-            elementToAdjust->updateEffectiveDirectionalityOfDirAuto();
-            return;
+    if (RefPtr shadowRoot = element.shadowRoot()) {
+        for (Ref child : childrenOfType<Element>(*shadowRoot)) {
+            if (child.ptr() == exclude)
+                continue;
+
+            updateElementAndShadowEffectiveTextDirection(child, direction);
+            child->updateEffectiveTextDirectionOfDescendants(direction);
         }
     }
+
+    if (element.renderer() && element.renderer()->style().direction() != effectiveDirection)
+        element.invalidateStyleForSubtree();
 }
 
-void Element::updateTextDirectionalityAfterInputTypeChange()
+static void updateElementTextDirection(Element& element, TextDirectionState textDirectionState, Element* exclude = nullptr)
```

Comment:
Maybe we should call this `updateEffectiveTextDirection`?

Reply:
I called it `updateEffectiveTextDirectionState()` since it takes `TextDirectionState`. This is to differentiate it from the functions which take `std::optional<TextDirection>`